Add static_assert to spanReinterpretCast<>() to prevent unnecessary casts

ddkilzer opened the pull request:
#### 4543cc7cb73c575744c415cdb70d9f7353e3181c
<pre>
Add static_assert to spanReinterpretCast&lt;&gt;() to prevent unnecessary casts
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=281994">https://bugs.webkit.org/show_bug.cgi?id=281994</a>&gt;
&lt;<a href="https://rdar.apple.com/138491814">rdar://138491814</a>&gt;

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Add static_assert() to catch unnecessary casts.
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeSpan):
- Eliminate unnecessary call to spanReinterpretCast&lt;&gt;() when decoding to
  uint8_t bytes.

Canonical link: <a href="https://commits.webkit.org/285623@main">https://commits.webkit.org/285623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d593cf363a99919934d8682ab9805d4b0d31a423

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73277 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26085 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77507 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24515 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/489 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57581 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76344 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63055 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38000 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44245 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20535 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22844 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66412 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66099 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20888 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79159 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72534 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/135 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63065 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7295 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94314 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11284 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/556 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3293 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20738 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/586 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/570 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/588 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
A little annoying that the only case this turned up was essentially a false positive. But I am still OK with this change.

Commit:
Add static_assert to spanReinterpretCast<>() to prevent unnecessary casts
<https://bugs.webkit.org/show_bug.cgi?id=281994>
<rdar://138491814>

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Add static_assert() to catch unnecessary casts.
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeSpan):
- Eliminate unnecessary call to spanReinterpretCast<>() when decoding to
  uint8_t bytes.

Canonical link: https://commits.webkit.org/285623@main

cdumez commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -764,14 +764,15 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
 constexpr std::span<T, Extent == std::dynamic_extent ? std::dynamic_extent : (sizeof(U) * Extent) / sizeof(T)> spanReinterpretCast(std::span<U, Extent> span)
 {
+    static_assert(std::is_const_v<T> || (!std::is_const_v<T> && !std::is_const_v<U>), "spanReinterpretCast will not remove constness from source");
+    static_assert(!std::is_same_v<std::remove_const_t<T>, std::remove_const_t<U>>, "Unnecessary call to spanReinterpretCast");
```
Personally, I do not like this. This makes span adoption harder because we have to write *more* code in templated functions like in Decoder.h to do the same thing (it's not just in Decoder.h, I'm hitting the exact same issue in another patch).

I'd be OK with that still if there were clear benefits. However, it is not clear to me the benefits are worth it here. There is no cost to an unnecessary reinterpret cast, is there?

darinadler replied:
My opinion only:

If this makes template code annoying to write a second time then I agree it could be removed. In templates it can be useful to be able to cast without being so careful about the "null case" where the types match.

But in general outside a template type situation a mysterious cast isn't good, so one that says `reinterpret_cast` but isn't really doing anything at all is a small negative thing, not because it does harm but because it can cause slight confusion.

So if the bad template thing happens only **once**, I'd slightly lean toward keeping it.

cdumez replied:
As I've mentioned, it's not once anymore. The reason I'm commenting is because I've just hit this during span adoption.

cdumez replied:
```
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:26:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/config.h:31:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/FastMalloc.h:26:
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/StdLibExtras.h:768:19: error: static assertion failed due to requirement '!std::is_same_v<unsigned char, unsigned char>': Unnecessary call to spanReinterpretCast
  768 |     static_assert(!std::is_same_v<std::remove_const_t<T>, std::remove_const_t<U>>, "Unnecessary call to spanReinterpretCast");
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:119:40: note: in instantiation of function template specialization 'WTF::spanReinterpretCast<unsigned char, unsigned char, 18446744073709551615UL>' requested here
  119 |     if (!decoder.decodeFixedLengthData(spanReinterpretCast<uint8_t>(buffer)))
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:143:16: note: in instantiation of function template specialization 'WTF::Persistence::decodeStringText<unsigned char>' requested here
  143 |         return decodeStringText<LChar>(decoder, *length);
1 error generated.
```

cdumez replied:
> But in general outside a template type situation a mysterious cast isn't good, so one that says reinterpret_cast but isn't really doing anything at all is a small negative thing, not because it does harm but because it can cause slight confusion.

I agree that unnecessary casts aren't good. However, I don't know that this is a problem we currently need to solve. We haven't found any case of unnecessary span casting outside template code yet. However, we've found 2 instances where we need to write more code to do achieve the same behavior, which is unfortunate.

As we're doing more and more span adoption for security reasons, I don't think we want to make that work more painful than it needs to be.

cdumez replied:
I've hit this a third time in decodeStringText() in DaemonCoders.h.

cdumez replied:
You can see those cases in https://github.com/WebKit/WebKit/pull/35793. For now, I'm proposing to revert this new check (which is what I'm currently doing as part of my PR). I can be convinced to go the hard way and add constexpr is_same_v checks if people really prefer that but it's not my preference.

darinadler replied:
Sure, I agree, let’s take it this check out. It was my suggestion and you have convinced me!
